resource_elasticsearch_xpack_user.go

```diff
@@ -0,0 +1,294 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+
+	elastic7 "github.com/olivere/elastic/v7"
+	elastic5 "gopkg.in/olivere/elastic.v5"
+	elastic6 "gopkg.in/olivere/elastic.v6"
+)
+
+func resourceElasticsearchXpackUser() *schema.Resource {
+
+	return &schema.Resource{
+		Create: resourceElasticsearchXpackUserCreateOrUpdate,
+		Read:   resourceElasticsearchXpackUserRead,
+		Update: resourceElasticsearchXpackUserCreateOrUpdate,
+		Delete: resourceElasticsearchXpackUserDelete,
+		Schema: map[string]*schema.Schema{
+			"username": &schema.Schema{
+				Type:         schema.TypeString,
+				Description:  "A unique identifier for the user",
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 1024),
+			},
+			"email": &schema.Schema{
+				Type:        schema.TypeString,
+				Description: "The email of the user",
+				Optional:    true,
+			},
+			"full_name": &schema.Schema{
+				Type:        schema.TypeString,
+				Description: "The full name of the user",
+				Optional:    true,
+			},
+			"enabled": &schema.Schema{
+				Type:        schema.TypeBool,
+				Description: "Specifies whether the user is enabled. The default value is true.",
+				Optional:    true,
+				Default:     true,
+			},
+			"password": &schema.Schema{
+				Type:         schema.TypeString,
+				Description:  "The user’s password. Passwords must be at least 6 characters long.",
+				Optional:     true,
+				Sensitive:    true,
+				ValidateFunc: validation.StringLenBetween(6, 1024),
+			},
+			"password_hash": &schema.Schema{
+				Type:        schema.TypeString,
+				Description: "A hash of the user’s password. This must be produced using the same hashing algorithm as has been configured for password storage",
```

Comment:
If this is optional, then what's it for?

Reply:
it is optional to specify either the `password_hash` or `password`, but not both. You will receive an error if you specify both.

Reply:
Also line length nitpick here. I really have a hard time reading lines that are longer than 100 chars.
```suggestion
				Description: `A hash of the user’s password. This must be produced using the same hashing
				algorithm as has been configured for password storage`,
```

Reply:
👍

Reply:
I tried writing a test using the password hash but couldn't get it working. I used this `docker-compose run opendistro /bin/bash -c plugins/opendistro_security/tools/hash.sh` command to generate a password hash from a password. Still complained that the password hash was using an incorrect hash algorithm: 

```
elastic: Error 400 (Bad Request): Provided password hash uses [NOOP] but the configured hashing algorithm is [BCRYPT] [type=illegal_argument_exception]
```